internal/langserver/handlers/initialize.go

```diff
@@ -188,8 +105,150 @@ func (svc *service) Initialize(ctx context.Context, params lsp.InitializeParams)
 		})
 	}
 
+	if params.RootURI == "" {
+		svc.singleFileMode = true
+		properties["root_uri"] = "file"
+		// return serverCaps, fmt.Errorf("Editing a single file is not yet supported." +
+		// 	" Please open a directory.")
+		jrpc2.ServerFromContext(ctx).Notify(ctx, "window/showMessage", &lsp.ShowMessageParams{
+			Type:    lsp.Warning,
+			Message: "Some of VS Code's capabilities are reduced when editing a single file, but you can still open text files and edit them. Please open a directory for full functionality.",
```

Comment:
```suggestion
			Message: "Some capabilities may be reduced when editing a single file. We recommend opening a directory for full functionality.",
```
I don't think we can imply any particular client here

Reply:
Applied edit